File: IdCreater.php
<?php
/**
 * Created by PhpStorm.
 * User: chenyuan3
 * Date: 2018/4/10
 * Time: 下午2:32
 */

namespace common;
for ($i = 0; $i < 10000; $i++) {
    $id = IdCreater::intiCreater()->getId();
    echo $id . "\n";
}


class IdCreater
{
    /**
     * id create 的诞生时间 用于 生成唯一 id
     *
     * @var int
     */
    const BIRTH_DAY = 1523342555743;

    /**
     * 服务器的编码所占字节 （1023）
     *
     * @var int
     */
    const WORKER_ID_BIT = 10;

    /**
     * 毫秒内生成重复id时解决碰撞的拉链 字节 （4095）
     *
     * @var int
     */
    const ZIPPER_BIT = 12;

    /**
     * 程序出现异常返回这个id
     */
    const DEFAULT_ID = 99999999999999;

    /**
     * id生成器实例
     *
     * @var \common\IdCreater
     */
    private static $CREATER;

    /**
     * 服务器id
     *
     * @var int
     */
    private static $WORK_ID;

    /**
     * 最近一次生成id的时间戳
     *
     * @var int
     */
    private static $LAST_TIMESTAMP;

    /**
     * 拉链的个数
     *
     * @var int
     */
    private static $ZIPPER_NUM = 0;

    /**
     * IdCreater constructor.
     *
     * @throws \Exception
     */
    private function __construct()
    {
        self::$WORK_ID = self::getWorkId();
        self::$LAST_TIMESTAMP = self::getTimestamp();
    }

    public static function intiCreater()
    {
        if (empty(self::$CREATER)) {
            self::$CREATER = new self();
        }

        return self::$CREATER;
    }

    /**
     * 生成id
     *
     * @return int
     */
    public function getId()
    {
        $timestamp = self::getTimestamp();
        if ($timestamp < self::$LAST_TIMESTAMP) {
            return self::DEFAULT_ID;
        }

        $max = pow(2, self::ZIPPER_BIT) - 1;
        if ($timestamp == self::$LAST_TIMESTAMP && $max > self::$ZIPPER_NUM) {
            self::$ZIPPER_NUM = self::$ZIPPER_NUM + 1;
        } else {
            /**
             * 当时间戳发生变化是改变
             */
            self::$LAST_TIMESTAMP = self::getNextTimestamp();
            self::$ZIPPER_NUM = 0;
        }

        return self::makeId();
    }

    /**
     * @return int
     */
    private static function getWorkId()
    {
//        if ($id > (pow(2, self::WORKER_ID_BIT) - 1)) {
//        }

        return 1;
    }

    private static function getTimestamp()
    {
        return intval(microtime(true) * 1000);
    }

    private static function getNextTimestamp()
    {
        $timestamp = self::getTimestamp();
        while ($timestamp <= self::$LAST_TIMESTAMP) {
            $timestamp = self::getTimestamp();
        }

        return $timestamp;
    }

    /**
     * 制作id
     * @return int
     */
    private static function makeId() {

        //时间戳向左移动的位数
        $timestampLeftMoveBit = self::ZIPPER_BIT + self::WORKER_ID_BIT;
        //服务器id向左移动的位数
        $workerIdLeftMoveBit = self::WORKER_ID_BIT;

        /**
         * 将位移运算转化成可读的数学运算， 方便理解
         */
        $first = (self::$LAST_TIMESTAMP - self::BIRTH_DAY) *  pow(2,$timestampLeftMoveBit);
        $second = self::$WORK_ID * pow(2,$workerIdLeftMoveBit);
        $third = self::$ZIPPER_NUM;

        /**
         * 等价于这个位移预算
         * (self::$LAST_TIMESTAMP - self::BIRTH_DAY) << $timestampLeftMoveBit | self::$WORK_ID << $workerIdLeftMoveBit | self::$ZIPPER_NUM;
         */
        return $first + $second + $third;
    }
}

File: IdCreater.lua
IdCreater = {workId =0, lastTimestamp=0, zipperNum=0}

IdCreater.BIRTH_DAY = 1523342555743
IdCreater.WORKER_ID_BIT = 10
IdCreater.ZIPPER_BIT = 12
IdCreater.DEFAULT_ID = 99999999999999
IdCreater.ENV_PATH  = '/data1/sinawap/code/run/application/config/.env'

function IdCreater:new()
    creater = {}
    setmetatable(creater, self)
    self.__index = self
    self.workId = self.getWorkId()
    self.lastTimestamp = self.getTimestamp()
    
    return creater
end

function IdCreater:getId() 
	local timeStamp = self.getTimestamp()

	if timeStamp < self.lastTimestamp
	then
		return IdCreater.DEFAULT_ID
	end

	local maxZipperNumer = math.pow(2, self.ZIPPER_BIT) - 1

	if timeStamp == self.lastTimestamp and maxZipperNumer > self.zipperNum
	then
		self.zipperNum = self.zipperNum + 1
	else
		self.lastTimestamp = self.getNextTimestamp()
		self.zipperNum = 0
	end

	return self:makeId()
end

function IdCreater:makeId()
	local timestampLeftMoveBit = self.ZIPPER_BIT + self.WORKER_ID_BIT
    local workerIdLeftMoveBit = self.WORKER_ID_BIT

    local first  = math.ceil((self.lastTimestamp - IdCreater.BIRTH_DAY) *  math.pow(2, timestampLeftMoveBit))
    local second = math.ceil(self.workId *  math.pow(2, workerIdLeftMoveBit))
    local third  = self.zipperNum

    return string.format("%d", first + second + third)
end

function IdCreater:getNextTimestamp()
    local timestamp = self.getTimestamp()

    while (self.lastTimestamp <= lastTimestamp) 
    do
        timestamp = self.getTimestamp()
    end

    return timestamp
end

function IdCreater:getWorkId()
    local file = io.open(IdCreater.ENV_PATH,"r")
    local workId = 1
    for line in file:lines() do
        _,_,workId=string.find(line,"SERVER_ID=(%d+)");  
    end
    if workId == nil or (tonumber(workId) > math.pow(2, IdCreater.WORKER_ID_BIT) - 1)
    then
        return 1
    end

    file.close()
	return tonumber(workId)
end

function IdCreater:getTimestamp()
    -- return os.time() * 1000
	return ngx.now() * 1000
end


function getId()
    idcreater = IdCreater:new(workId);
    return idcreater:getId()
end

function myerrorhandler( err )
   print( "ERROR:", err )
end

status, id = xpcall( getId, myerrorhandler)

print(id)

if status == false 
then
	return IdCreater.DEFAULT_ID
else
	return id
end





